pkg/apiserver/debugapi/service.go

```diff
@@ -101,18 +89,7 @@ func (s *Service) RequestEndpoint(c *gin.Context) {
 		return
 	}
 
-	ep, ok := s.endpointMap[req.ID]
-	if !ok {
-		utils.MakeInvalidRequestErrorWithMessage(c, "Invalid endpoint id: %s", req.ID)
-		return
-	}
-	endpointReq, err := ep.NewRequest(req.Host, req.Port, req.Params)
-	if err != nil {
-		_ = c.Error(err)
-		return
-	}
-
-	res, err := SendRequest(ep.Client, endpointReq)
+	res, err := s.Client.Send(req.ID, req.Host, req.Port, req.Params)
```

Comment:
How about rearranging the code as follows:

debugapi/endpoint/:

```go
type EndpointRequestPayload struct {
  EndpointID     string            `json:"endpoint_id"`
  Host   string            `json:"host"`
  Port   int               `json:"port"`
  Params map[string]string `json:"params"`
}

type EndpointRequestPayloadResolved struct {
  originalPayload EndpointRequestPayload
  timeout time.Duration
  params map[string]interface{}
}

func (api *APIModel) ResolvePayload(payload EndpointRequestPayload) (*EndpointRequestPayloadResolved, error) {
  resolved := &resolvedEndpointRequestPayload{
    originalPayload: payload,
    params: new(...),
  }
  // 1. execute `api.OnResolve(resolved)` to update the request the timeout.
  // 2. execute each parameter's `OnResolve(resolved)` to verify payload and update the `resolved.params`
  return resolved, nil
}

func (xx) SendToEndpoint(payload EndpointRequestPayload) .. {
  // 1. resolve the endpoint via the endpoint_id in payload
  // 2. 
  resolvedPayload := endpoint.Resolve(payload)
  // 3. Send to corresponding client with the `resolvedPayload`.
}
```

I believe this makes your code shorter, since there are just less concepts involved.